### PR TITLE
TUM-06 - General project design from BE perspective.

### PR DESCRIPTION
**Pull request by Camed**
General BE project structure. Still a lot of things to implement and connect to  presentation layer.

### CODE REVIEW

File: Backend/Tumultu/Application/Common/Interfaces/IIdentityService.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using Tumultu.Domain.Entities;
+
+namespace Tumultu.Application.Common.Interfaces;
+
+public interface IIdentityService
+{
+    Task<User?> GetUserById(Guid id);
+    Task<bool> IsInRoleAsync(Guid? id, Roles role);
+    Task<bool> AuthorizeAsync(Guid? Id, Roles role);
+    Task<(Result result, Guid userId)> CreateUserAsync(string userName, string password);
```
Comment (PaterMajkel) [inline Backend/Tumultu/Application/Common/Interfaces/IIdentityService.cs]: Can you make a specific model for that? Don't leave it as a genric type, please

File: Backend/Tumultu/Application/Common/Models/Result.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿namespace Tumultu.Application.Common.Models;
+
+public class Result
+{
+    internal Result(bool success, IEnumerable<string> errors)
```
Comment (PaterMajkel) [inline Backend/Tumultu/Application/Common/Models/Result.cs:5]: Should it be internal?

File: Backend/Tumultu/Application/Files/Commands/CreateFileCommandValidator.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+public class CreateFileCommandValidator : AbstractValidator<CreateFileCommand>
+{
+    public CreateFileCommandValidator()
+    {
+        RuleFor(x => x.Payload)
+            .NotNull()
+            .NotEmpty();
+
+        RuleFor(x => x.Payload.Length)
+            .LessThanOrEqualTo(26_214_400) // 25mb max
```
Comment (PaterMajkel) [inline Backend/Tumultu/Application/Files/Commands/CreateFileCommandValidator.cs]: We can pass these numbers to some config file to make future changes easier

File: Backend/Tumultu/Domain/Common/ValueObject.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/// <summary>
+/// Value object created with Microsoft guidelines.
+/// Find more here: https://learn.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/implement-value-objects
+/// </summary>
+public abstract class ValueObject
+{
+    protected abstract IEnumerable<object> GetEqualityComponents();
+    protected static bool EqualOperator(ValueObject l, ValueObject r)
+    {
+        if(l is null ^ r is null)
```
Comment (PaterMajkel) [inline Backend/Tumultu/Domain/Common/ValueObject.cs:20]: very cool, can you add a comment that it's a XOR operator? Didn't know that and I'm certain that others might also not know that. Might save a search further down the line xd

File: Backend/Tumultu/Infrastructure/Infrastructure.csproj [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (PaterMajkel) [inline Backend/Tumultu/Infrastructure/Infrastructure.csproj]: How about .net 8? xd 👉👈

File: Backend/Tumultu/Presentation/Program.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+var builder = WebApplication.CreateBuilder(args);
+
+// Add services to the container.
+
+builder.Services.AddControllers();
+// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
```
Comment (PaterMajkel) [inline Backend/Tumultu/Presentation/Program.cs:8]: Can you change it to openAPI swash? It will help with migrating models to anular

**Review by PaterMajkel — APPROVED** [review]
Made some suggestions but it looks good

File: Backend/Tumultu/Application/Common/Models/Result.cs (hunk 2 above)
Comment (Camed) [reply Backend/Tumultu/Application/Common/Models/Result.cs:5]: Yes 👍

File: Backend/Tumultu/Application/Files/Commands/CreateFileCommandValidator.cs (hunk 3 above)
Comment (Camed) [reply Backend/Tumultu/Application/Files/Commands/CreateFileCommandValidator.cs]: Agreed, on it

File: Backend/Tumultu/Infrastructure/Infrastructure.csproj (hunk 5 above)
Comment (Camed) [reply Backend/Tumultu/Infrastructure/Infrastructure.csproj]: When I was starting project it was still unavailable for me, will change soon.

File: Backend/Tumultu/Domain/Common/ValueObject.cs (hunk 4 above)
Comment (Camed) [reply Backend/Tumultu/Domain/Common/ValueObject.cs:20]: I thought it is kinda basic knowledge, but sure, I can add comment xD
This file is generally copypaste from MS docs so

File: Backend/Tumultu/Presentation/Program.cs (hunk 6 above)
Comment (Camed) [reply Backend/Tumultu/Presentation/Program.cs:8]: To be honest I've never done that. Will look into it

File: Backend/Tumultu/Application/Common/Interfaces/IIdentityService.cs (hunk 1 above)
Comment (Camed) [reply Backend/Tumultu/Application/Common/Interfaces/IIdentityService.cs]: You mean only tuple, line 12?

**Pull request merged**